File: app/src/main/java/com/example/practiceproject/MetroService.java
package com.example.practiceproject;

import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.List;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Used Bill Pugh implementation of the singleton pattern
 */
public class MetroService {

    private static final String BASE_URL = "https://my-json-server.typicode.com/BeeWhy/metro/";

    private MetroService() {
    }

    private static final class InstanceHolder {
        static final MetroService INSTANCE = new MetroService();
    }

    public static MetroService getInstance() {
        return InstanceHolder.INSTANCE;
    }

    public void addMetroMarkers(GoogleMap map) {
        Call<List<Station>> stations = getStations();
        stations.enqueue(new Callback<List<Station>>() {
            @Override
            public void onResponse(@NonNull Call<List<Station>> call,
                                   @NonNull Response<List<Station>> response) {
                if (response.body() != null) {
                    for (Station station : response.body()) {
                        if (!Objects.equals(station.getName().toLowerCase(), "error")) {
                            map.addMarker(new MarkerOptions()
                                    .position(new LatLng(station.getLatitude(), station.getLongitude()))
                                    .title(station.getName()));
                        }
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<Station>> call,
                                  @NonNull Throwable t) {
                Log.e(MetroService.class.getSimpleName(), "Response fail", t);
            }
        });
    }

    private Call<List<Station>> getStations() throws SecurityException{
        if (PermissionsUtils.isInternetPermissionGranted()) {
            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
            MetroAPI metroAPI = retrofit.create(MetroAPI.class);
            return metroAPI.stations();
        } else {
            throw new SecurityException();
        }
    }
}

File: app/src/main/java/com/example/practiceproject/MetroService.kt
package com.example.practiceproject

import android.util.Log
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.MarkerOptions
import kotlin.Throws
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.bottomsheet.BottomSheetBehavior
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

/**
 * Used singleton pattern
 */
object MetroService {
    private const val BASE_URL = "https://my-json-server.typicode.com/BeeWhy/metro/"

    fun addMetroMarkers(map: GoogleMap) {
        val stations = stations
        stations.enqueue(object : Callback<List<Station>?> {
            override fun onResponse(
                call: Call<List<Station>?>,
                response: Response<List<Station>?>
            ) {
                if (response.body() != null) {
                    for (station in response.body()!!) {
                        if (station.name.lowercase(Locale.getDefault()) != "error") {
                            map.addMarker(
                                MarkerOptions()
                                    .position(
                                        LatLng(
                                            station.latitude.toDouble(),
                                            station.longitude.toDouble()
                                        )
                                    )
                                    .title(station.name)
                            )
                        }
                    }
                }
            }

            override fun onFailure(
                call: Call<List<Station>?>,
                t: Throwable
            ) {
                Log.e(MetroService::class.java.simpleName, "Response fail", t)
            }
        })
    }

    fun addStationsList(layout : LinearLayout, behavior : BottomSheetBehavior<LinearLayout>) {
        stations.enqueue(object : Callback<List<Station>?> {
            override fun onResponse(
                call: Call<List<Station>?>,
                response: Response<List<Station>?>
            ) {
                if (response.body() != null) {
                    for (station in response.body()!!) {
                        if (station.name.lowercase(Locale.getDefault()) != "error") {
                            val textView = TextView(layout.context)
                            textView.text = station.name
                            textView.setOnClickListener { view: View ->
                                if (behavior.state == BottomSheetBehavior.STATE_COLLAPSED
                                    && (view as TextView).text == textView.text
                                ) {
                                    behavior.state = BottomSheetBehavior.STATE_HIDDEN
                                } else {
                                    behavior.state = BottomSheetBehavior.STATE_COLLAPSED
                                }
                            }
                            layout.addView(textView)
                        }
                    }
                }
            }

            override fun onFailure(call: Call<List<Station>?>, t: Throwable) {
                t.printStackTrace()
            }
        })
    }

    @get:Throws(SecurityException::class)
    private val stations: Call<List<Station>>
        get() = if (PermissionsUtils.isInternetPermissionGranted) {
            val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
            val metroAPI = retrofit.create(MetroAPI::class.java)
            metroAPI.stations()
        } else {
            throw SecurityException()
        }
}